push/sender/src/main/java/org/jboss/aerogear/unifiedpush/message/MetricsCollector.java

```diff
@@ -16,40 +16,69 @@
  */
 package org.jboss.aerogear.unifiedpush.message;
 
+import javax.annotation.Resource;
 import javax.ejb.Stateless;
+import javax.enterprise.event.Event;
 import javax.enterprise.event.Observes;
 import javax.inject.Inject;
+import javax.jms.Queue;
 
 import org.jboss.aerogear.unifiedpush.api.PushMessageInformation;
 import org.jboss.aerogear.unifiedpush.api.VariantMetricInformation;
+import org.jboss.aerogear.unifiedpush.message.holder.PushMessageCompleted;
+import org.jboss.aerogear.unifiedpush.message.holder.VariantCompleted;
+import org.jboss.aerogear.unifiedpush.message.jms.AbstractJMSMessageConsumer;
 import org.jboss.aerogear.unifiedpush.message.jms.Dequeue;
 import org.jboss.aerogear.unifiedpush.service.metrics.PushMessageMetricsService;
+import org.jboss.aerogear.unifiedpush.utils.AeroGearLogger;
 
 /**
  * Receives metrics from {@link NotificationDispatcher} and updates the database.
  */
 @Stateless
-public class MetricsCollector {
+public class MetricsCollector extends AbstractJMSMessageConsumer {
+
+    private final AeroGearLogger logger = AeroGearLogger.getInstance(MetricsCollector.class);
 
     @Inject
     private PushMessageMetricsService metricsService;
 
+    @Resource(mappedName = "java:/queue/BatchLoadedQueue")
+    private Queue batchLoadedQueue;
+
+    @Resource(mappedName = "java:/queue/AllBatchesLoadedQueue")
+    private Queue allBatchesLoaded;
+
+    @Inject
+    private Event<VariantCompleted> variantCompleted;
+
+    @Inject
+    private Event<PushMessageCompleted> pushMessageCompleted;
+
     /**
      * Receives metrics and update the push message information in database.
      *
      * @param variantMetricInformation the variant metrics info object
      */
     public void collectMetrics(@Observes @Dequeue VariantMetricInformation variantMetricInformation) {
-        PushMessageInformation pushMessageInformation = variantMetricInformation.getPushMessageInformation();
+        PushMessageInformation pushMessageInformation = metricsService.getPushMessageInformation(variantMetricInformation.getPushMessageInformation().getId());
+        metricsService.lock(pushMessageInformation);
```

Comment:
Nope, we had concurrency issues here - if you send really small transactions or send them quickly, lot of `VariantMetricInformation` are coming concurrently.

Two `MetricsCollector`s were able to work concurrently on one `PushMessageInformation`, thus rewriting their updates.

We could do optimistic locking with `@Version` though, should I open a JIRA? @matzew didn't want to introduce that now, but maybe later?

If lock fails / times out, the transaction is aborted and `VariantMetricInformation` is delivered later.